src/shims/x86/mod.rs

```diff
@@ -286,3 +298,49 @@ fn bin_op_simd_float_all<'tcx, F: rustc_apfloat::Float>(
 
     Ok(())
 }
+
+/// Horizontaly performs `which` operation on adjacent values of
+/// `left` and `right` SIMD vectors and stores the result in `dest`.
+fn horizontal_bin_op<'tcx>(
+    this: &mut crate::MiriInterpCx<'_, 'tcx>,
+    which: mir::BinOp,
+    saturating: bool,
+    left: &OpTy<'tcx, Provenance>,
+    right: &OpTy<'tcx, Provenance>,
+    dest: &PlaceTy<'tcx, Provenance>,
+) -> InterpResult<'tcx, ()> {
+    let (left, left_len) = this.operand_to_simd(left)?;
+    let (right, right_len) = this.operand_to_simd(right)?;
+    let (dest, dest_len) = this.place_to_simd(dest)?;
+
+    assert_eq!(dest_len, left_len);
+    assert_eq!(dest_len, right_len);
+    assert_eq!(dest_len % 2, 0);
+
+    let middle = dest_len / 2;
+    for i in 0..dest_len {
+        let (lhs, rhs) = if i < middle {
+            let base_i = i.checked_mul(2).unwrap();
+            (
+                this.read_immediate(&this.project_index(&left, base_i)?)?,
+                this.read_immediate(&this.project_index(&left, base_i.checked_add(1).unwrap())?)?,
+            )
+        } else {
+            let base_i = i.checked_sub(middle).unwrap().checked_mul(2).unwrap();
+            (
+                this.read_immediate(&this.project_index(&right, base_i)?)?,
+                this.read_immediate(&this.project_index(&right, base_i.checked_add(1).unwrap())?)?,
+            )
+        };
```

Comment:
The actual difference here is whether we apply `.checked_sub(middle)` and whether we use `left` or `right` as our source vector. Can this be rewritten along those lines? I'd rather read something like this because it's more clear what's different where:
```rust
let (i, src) = if i < middle {
    (i, &left)
} else {
    (i.checked_sub(middle).unwrap(), &*right)
};
let base_i = i .checked_mul(2).unwrap();
let lhs = this.read_immediate...
```

Reply:
Yeah, that's definitely better.